crates/biome_css_analyze/src/lint/nursery/no_duplicate_custom_properties.rs

```diff
@@ -0,0 +1,94 @@
+use biome_analyze::{context::RuleContext, declare_rule, Ast, Rule, RuleDiagnostic};
+use biome_console::markup;
+use biome_css_syntax::{CssDeclarationOrRuleList, CssLanguage};
+use biome_rowan::{SyntaxToken, TextRange};
+
+declare_rule! {
+    /// Disallow duplicate custom properties within declaration blocks.
+    ///
+    /// This rule checks the declaration blocks for duplicate custom properties.
+    ///
+    /// ## Examples
+    ///
+    /// ### Invalid
+    ///
+    /// ```css,expect_diagnostic
+    /// a { --custom-property: pink; --custom-property: orange;  }
+    /// ```
+    ///
+    /// ```css,expect_diagnostic
+    /// a { --custom-property: pink; background: orange; --custom-property: orange }
+    /// ```
+    ///
+    /// ### Valid
+    ///
+    /// ```css
+    /// a { --custom-property: pink; }
+    /// ```
+    ///
+    /// ```css
+    /// a { --custom-property: pink; --cUstOm-prOpErtY: orange; }
+    /// ```
+    ///
+    pub NoDuplicateCustomProperties {
+        version: "next",
+        name: "noDuplicateCustomProperties",
+        recommended: false,
+    }
+}
+
+impl Rule for NoDuplicateCustomProperties {
+    type Query = Ast<CssDeclarationOrRuleList>;
+    type State = TextRange;
+    type Signals = Option<Self::State>;
+    type Options = ();
+
+    fn run(ctx: &RuleContext<Self>) -> Option<Self::State> {
+        let node = ctx.query();
+        let mut custom_properties: Vec<SyntaxToken<CssLanguage>> = vec![];
+        for item in node {
+            if let Some(css_declaration) = item.as_css_declaration_with_semicolon() {
+                if let Ok(property) = css_declaration.declaration().ok()?.property() {
+                    if let Some(property) = property.as_css_generic_property() {
+                        if let Some(ident) = property.name().ok()?.as_css_identifier() {
+                            let value_token = ident.value_token().ok()?;
+                            custom_properties.push(value_token);
+                        }
+                    }
+                }
+            }
+        }
```

Comment:
I think you can get rid of the `mut` here by using a `filter_map` and it will return an `Iterator` that you can possible use directly with `check_duplicate_custom_properties`.

```suggestion
        let custom_properties = node.into_iter().filter_map(|item| {
            item.as_css_declaration_with_semicolon()
                .and_then(|css_declaration| {
                    css_declaration
                        .declaration()
                        .ok()?
                        .property()
                        .ok()?
                        .as_css_generic_property()
                        .and_then(|css_generic_property| {
                            css_generic_property
                                .name()
                                .ok()?
                                .as_css_identifier()
                                .and_then(|css_identifier| css_identifier.value_token().ok())
                        })
                })
        });
```

You don't need to use it like above, as long as you are returning an `Option` (`Some`/`None`) it should be okay!

Reply:
@abidjappie 
I appreciate so much for your feedback!! That's clean and readable.